### PR TITLE
feat: public preview support for microsoft.azuremonitor.containers.metrics in ARC clusters (managed prometheus)

### PR DESCRIPTION
**Pull request by bragi92**
---

This checklist is used to make sure that common guidelines for a pull request are followed.

### Related command

`az k8s-extension create --name azuremonitor-metrics -g kaveesharc -c kaveesharc -t connectedClusters --extension-type Microsoft.AzureMonitor.Containers.Metrics --configuration-settings azure-monitor-workspace-resource-id="{amw_id}" AzureMonitorMetrics.KubeStateMetrics.MetricAnnotationsAllowList="pods=[k8s-annotation-1,k8s-annotation-n]"`

`az k8s-extension delete --name azuremonitor-metrics -g kaveesharc -c kaveesharc -t connectedClusters`

Here is a link to the Managed Cluster PR : https://github.com/Azure/azure-cli/pull/26201

### General Guidelines

- [ ] Have you run `azdev style <YOUR_EXT>` locally? (`pip install azdev` required)
- [ ] Have you run `python scripts/ci/test_index.py -q` locally?

For new extensions:

- [ ] My extension description/summary conforms to the [Extension Summary Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/extensions/extension_summary_guidelines.md).

### About Extension Publish

There is a pipeline to automatically build, upload and publish extension wheels.  
Once your pull request is merged into main branch, a new pull request will be created to update `src/index.json` automatically.  
You only need to update the version information in file setup.py and historical information in file HISTORY.rst in your PR but do not modify `src/index.json`.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -88,6 +88,7 @@ venv/
 ENV/
 env/
 env27/
+azenv/
```
Comment (NarayanThiru) [inline .gitignore:91]: >azenv [](http://example.com/codeflow?start=0&length=5)

Why is this needed?  Can you revert this change?

File: src/k8s-extension/HISTORY.rst [hunk 2]
```diff
@@ -6,6 +6,7 @@ Release History
 1.4.1
 ++++++++++++++++++
 * microsoft.azureml.kubernetes: Fix sslSecret parameter in update operation
+* microsoft.azuremonitor.containers.metrics : public preview support for managed prometheus in ARC clusters
```
Comment (NarayanThiru) [inline src/k8s-extension/HISTORY.rst:10]: Create a new version 1.4.2 and add this comment

File: src/k8s-extension/setup.py [hunk 3]
```diff
@@ -33,7 +33,7 @@
 # TODO: Add any additional SDK dependencies here
 DEPENDENCIES = []
 
-VERSION = "1.4.0"
+VERSION = "1.4.1"
```
Comment (NarayanThiru) [inline src/k8s-extension/setup.py:36]: Looks like the previous update did not increment the version - fyi @bavneetsingh16 .
@bragi92, pl. change this to 1.4.2, which should have been the version for this update

File: src/k8s-extension/azext_k8s_extension/partner_extensions/azuremonitormetrics/azuremonitorprofile.py [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+        cluster_name,
+        configuration_settings,
+        cluster_type
+):
+    cloud_name = cmd.cli_ctx.cloud.name
+    if cloud_name.lower() == 'azurechinacloud':
+        raise CLIError("Azure China Cloud is not supported for the Azure Monitor Metrics extension")
+
+    if cloud_name.lower() == "azureusgovernment":
+        if configuration_settings is not None and 'grafana-resource-id' in configuration_settings:
```
Comment (NarayanThiru) [inline src/k8s-extension/azext_k8s_extension/partner_extensions/azuremonitormetrics/azuremonitorprofile.py:71]: >'grafana-resource-id' [](http://example.com/codeflow?start=50&length=21)

Is this case sensitive check?  IOW, if this key was input in different case, will this still work?

File: src/k8s-extension/azext_k8s_extension/partner_extensions/azuremonitormetrics/helper.py [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+            isInsightsRpRegistered = True
+        if value["namespace"].lower() == "microsoft.alertsmanagement" and value["registrationState"].lower() == "registered":
+            isAlertsManagementRpRegistered = True
+        if value["namespace"].lower() == "microsoft.monitor" and value["registrationState"].lower() == "registered":
+            isAlertsManagementRpRegistered = True
+        if value["namespace"].lower() == "microsoft.dashboard" and value["registrationState"].lower() == "registered":
+            isAlertsManagementRpRegistered = True
+    if isInsightsRpRegistered is False:
+        headers = ['User-Agent=arc-azuremonitormetrics.register_insights_rp']
+        post_request(cmd, subscription_id, "microsoft.insights", headers)
```
Comment (NarayanThiru) [inline src/k8s-extension/azext_k8s_extension/partner_extensions/azuremonitormetrics/helper.py:75]: >post_request(cmd, subscription_id, "microsoft.insights", headers) [](http://example.com/codeflow?start=8&length=65)

Since you are doing this without the customer knowing, it'd be good to print a message that this registration is being done.
Moreover, if the customer does not have sufficient permission to register this RP for that Sub, this will fail.  At that time, the message above will help explain the context

File: src/k8s-extension/azext_k8s_extension/partner_extensions/azuremonitormetrics/helper.py [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+def get_cluster_region(cmd, cluster_rp, subscription_id, cluster_resource_group_name, cluster_name, cluster_type):
+    cluster_region = ''
+    resources = cf_resources(cmd.cli_ctx, subscription_id)
+    cluster_resource_id = '/subscriptions/{0}/resourceGroups/{1}/providers/{2}/{3}/{4}'.format(
+        subscription_id, cluster_resource_group_name, cluster_rp, cluster_type, cluster_name)
+    try:
+        if cluster_rp.lower() == consts.HYBRIDCONTAINERSERVICE_RP:
+            resource = resources.get_by_id(cluster_resource_id, consts.HYBRIDCONTAINERSERVICE_API_VERSION)
+        else:
+            resource = resources.get_by_id(cluster_resource_id, '2020-01-01-preview')
```
Comment (NarayanThiru) [inline src/k8s-extension/azext_k8s_extension/partner_extensions/azuremonitormetrics/helper.py:96]: >'2020-01-01-preview' [](http://example.com/codeflow?start=64&length=20)

Move this to constants.py

File: src/k8s-extension/azext_k8s_extension/partner_extensions/azuremonitormetrics/helper.py [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+        post_request(cmd, subscription_id, "microsoft.monitor", headers)
+    if isDashboardRpRegistered is False:
+        headers = ['User-Agent=arc-azuremonitormetrics.register_dashboard_rp']
+        post_request(cmd, subscription_id, "microsoft.dashboard", headers)
+
+
+def get_cluster_region(cmd, cluster_rp, subscription_id, cluster_resource_group_name, cluster_name, cluster_type):
+    cluster_region = ''
+    resources = cf_resources(cmd.cli_ctx, subscription_id)
+    cluster_resource_id = '/subscriptions/{0}/resourceGroups/{1}/providers/{2}/{3}/{4}'.format(
```
Comment (NarayanThiru) [inline src/k8s-extension/azext_k8s_extension/partner_extensions/azuremonitormetrics/helper.py:90]: >/subscriptions/{0}/resourceGroups/{1}/providers/{2}/{3}/{4} [](http://example.com/codeflow?start=27&length=59)

Consider moving these Uri templates to constants too

File: src/k8s-extension/HISTORY.rst (hunk 2 above)
Comment (bavneetsingh16) [reply src/k8s-extension/HISTORY.rst:10]: Its okay to add it in the version '1.4.1' itself as the most recent version is 1.4.0 . The history was updated in another pr but was not released.

fyi @NarayanThiru, pr link --> [https://github.com/AzureArcForKubernetes/azure-cli-extensions/pull/224/files](https://github.com/AzureArcForKubernetes/azure-cli-extensions/pull/224/files)

File: src/k8s-extension/setup.py (hunk 3 above)
Comment (bavneetsingh16) [reply src/k8s-extension/setup.py:36]: 1.4.1 is good, it is not yet released

File: src/k8s-extension/azext_k8s_extension/partner_extensions/azuremonitormetrics/amg/link.py [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+            MonitoringDataReader
+        )
+        association_body = json.dumps({
+            "properties": {
+                "roleDefinitionId": roleDefinitionId,
+                "principalId": servicePrincipalId
+            }
+        })
+        headers = ['User-Agent=arc-azuremonitormetrics.add_role_assignment']
+        send_raw_request(cmd.cli_ctx, "PUT", roleDefinitionURI, body=association_body, headers=headers)
```
[the remark this answers is not in the document]
Comment (bragi92) [reply src/k8s-extension/azext_k8s_extension/partner_extensions/azuremonitormetrics/amg/link.py:56]: Yes, this is a valid scenario as it only happens when a customer is trying to link a grafana workspace to their AMW. If it fails, I have a error message being displayed which tells them what needs to be done to correct it but the overall onboarding will succeed as the extension installation will succeed without issue and it will even send data to the AMW.

[the remark this answers is not in the document]
Comment (NarayanThiru) [reply src/k8s-extension/azext_k8s_extension/partner_extensions/azuremonitormetrics/amg/link.py:56]: Thanks for explaining, Kaveesh!

**Review by NarayanThiru — DISMISSED** [review]
:shipit:

**Comment by bavneetsingh16** [comment]
can you please add test case for your extensions in this [folder](https://github.com/AzureArcForKubernetes/azure-cli-extensions/tree/k8s-extension/public/testing/test/extensions/public)

**Review by bavneetsingh16 — APPROVED** [review]
LGTM
